byzcoin/contracts/contract_deferred_test.go

```diff
@@ -2276,3 +2277,666 @@ func TestDeferred_ScenarioUpdateConfig(t *testing.T) {
+		Instructions: []byzcoin.Instruction{{
+			InstanceID: myID,
+			Invoke: &byzcoin.Invoke{
+				ContractID: byzcoin.ContractDeferredID,
+				Command:    "addProof",
+				Args: []byzcoin.Argument{
+					{
+						Name:  "identity",
+						Value: identityBuf,
+					},
+					{
+						Name:  "signature",
+						Value: signature,
+					},
+					{
+						Name:  "index",
+						Value: indexBuf,
+					},
+				},
+			},
+			SignerCounter: []uint64{3},
+		}},
+	}
+	require.Nil(t, ctx.FillSignersAndSignWith(signer))
+
+	_, err = cl.AddTransaction(ctx)
+	require.Nil(t, err)
+
+	proposedTransaction.Instructions[0].SignerIdentities = append(proposedTransaction.Instructions[0].SignerIdentities, identity)
+	proposedTransaction.Instructions[0].Signatures = append(proposedTransaction.Instructions[0].Signatures, signature)
+	result.ProposedTransaction = proposedTransaction
+
+	resultBuf, err := protobuf.Encode(&result)
+	require.Nil(t, err)
+
+	pr, err = cl.WaitProof(byzcoin.NewInstanceID(myID.Slice()), 2*genesisMsg.BlockInterval, resultBuf)
+	require.Nil(t, err)
+	require.True(t, pr.InclusionProof.Match(myID.Slice()))
+
+	dataBuf, _, _, err = pr.Get(myID.Slice())
+	require.Nil(t, err)
+
+	result = byzcoin.DeferredData{}
+	err = protobuf.Decode(dataBuf, &result)
+	require.Nil(t, err)
+
+	// We can not do this test because the identities have to be compared using
+	// the Equal() method.
+	//require.Equal(t, result.ProposedTransaction, proposedTransaction)
+	require.Equal(t, len(result.ProposedTransaction.Instructions), 1)
+	require.NotEmpty(t, result.ProposedTransaction.Instructions[0].SignerIdentities)
+	require.Equal(t, len(result.ProposedTransaction.Instructions[0].SignerIdentities), 1)
+	// This test won't work. But by using Equal() will.
+	// require.Equal(t, result.ProposedTransaction.Instructions[0].SignerIdentities[0], identity)
+	require.True(t, identity.Equal(&result.ProposedTransaction.Instructions[0].SignerIdentities[0]))
+
+	require.NotEmpty(t, result.ProposedTransaction.Instructions[0].Signatures)
+	require.Equal(t, len(result.ProposedTransaction.Instructions[0].Signatures), 1)
+	require.Equal(t, result.ProposedTransaction.Instructions[0].Signatures[0], signature)
+	// Default NumExecution should be 1
+	require.Equal(t, result.NumExecution, uint64(1))
+
+	require.NotEmpty(t, result.InstructionHashes)
+
+	local.WaitDone(genesisMsg.BlockInterval)
+
+	// ------------------------------------------------------------------------
+	// 4. Try to exec the proposed transaction. Should fail since only one
+	//    signer has signed.
+	// ------------------------------------------------------------------------
+
+	ctx = byzcoin.ClientTransaction{
+		Instructions: []byzcoin.Instruction{{
+			InstanceID: myID,
+			Invoke: &byzcoin.Invoke{
+				ContractID: byzcoin.ContractDeferredID,
+				Command:    "execProposedTx",
+			},
+			SignerCounter: []uint64{4},
+		}},
+	}
+	require.Nil(t, ctx.FillSignersAndSignWith(signer))
+
+	_, err = cl.AddTransaction(ctx)
```

Comment:
you can change a lot of `AddTransaction` to `AddTransactionAndWait(ctx, 10)` then it will only return when the transaction is included in the block. that way you don't need to do `time.Sleep(2 * genesisMsg.BlockInterval)` which may cause the test to run slower.

Reply:
Nice observation! Tests are now 2 times faster 🏆